### PR TITLE
[SCFToCalyx] Add support for index in- and output types

### PR DESCRIPTION
**Pull request by mortbopet**
This commit adds support for `index`-typed operands and return values.

### CODE REVIEW

File: test/Conversion/SCFToCalyx/convert_memory.mlir [hunk 1]
```diff
@@ -256,3 +256,74 @@ module {
     return %3 : i32
   }
 }
+
+// -----
+
+// Test index types as inputs
```
Comment (cgyurgyik) [inline test/Conversion/SCFToCalyx/convert_memory.mlir:262]: Periods please :-)

**Review by cgyurgyik — APPROVED** [review]
LGTM!

**Review by mikeurbach — APPROVED** [review]
Looks good to me, but I am curious if we should treat indices as `i64`. That's what we chose in the Handshake lowering. MLIR defines index types to have a platform-specific width, but internally represents them as 64 bits: https://github.com/llvm/llvm-project/blob/b12e4c17e07b42559fea219034dd162de25df498/mlir/include/mlir/IR/BuiltinTypes.td#L196. If we hit an index that is greater than what can be represented as `i32`, this could be an issue. Then again, perhaps we won't need to support that.

**Comment by mortbopet** [comment]
@mikeurbach bumping it up to 64 bits is an easy change, so if we find that we need/feel more secure doing so, then we should do it. Regardless, there should be a check to ensure that actual memory address widths are always narrower than our fixed width (we obviously don't want to infer `pad` operators). Then, this would catch the (unlikely) situation that a memory dimension is wider than 32/64 bits.

**Comment by mikeurbach** [comment]
Yep agreed. This is a good starting point.

**Pull request merged**